rfd/0086/multiprocess.md

```diff
@@ -117,54 +131,101 @@ ContainerPilot hasn't eliminated the complexity of dependency management -- that
 
-      # `app` needs this DB and also needs to configure itself when the DB
-      # is healthy. It reloads its config if the DB changes.
-      database:
-        wait: onHealthy
-        timeout: "60s"
-        onHealthy: configure-db-connnection.sh
-        onChange: reload-db-connections.sh
+The diagram below roughly describes the dependencies we have.
 
-      # `app` gracefully handles missing redis so we just update the config
-      # whenever the list of members changes
-      redis:
-        onChange: reload-configuration.sh
+```
+       +----> setup ------+
+       |                  |
+app ---+----> database ---+----> consul-agent ----> container start
+       |                  |
+       +~~~~> redis ------+
 ```
 
-Note that to avoid stalled containers, ContainerPilot will need to automatically detect unresolvable states. In the example above, if `setup` fails ContainerPilot will never start `app`, so it should mark the `app` service state as failed as well. If ContainerPilot reaches a state where no services can continue, it will exit.
+The configuration syntax for `start` doesn't permit multiple dependencies, but we can describe a chain of dependencies by forcing one of the hard dependencies to depend on the other as shown in the configuration below. The user could also accomplish the same dependency chain by merging the `configure-db-connection` and `setup` service executables, or even by having the `setup` executable poll the ContainerPilot status endpoint for more fine-grained control.
+
+```json5
+{
+  services: [
+    {
+      name: "consul-agent",
+      start: "startup", // this is the default
+      exec: "consul -agent -join {{ .CONSUL }}"
+    },
+    {
+      // the db is a hard dependency for 'app'
+      // onHealthy for an external db implies we are also waiting for
+      // consul-agent so we meet that requirement too
+      name: "configure-db-connection",
+      start: "onHealthy database",
+      exec: "/bin/configure-db.sh"
+    },
+    {
+      name: "setup",
+      start: "onSuccess configure-db-connection",
+      exec: "/bin/setup.sh"
+    },
+    {
+      name: "app",
+      // 'app' will never start if 'setup' fails
+      start: "onSuccess setup",
+      exec: "node /bin/app.js"
+    },
+  ],
+  health: [
+    {
+      // because we haven't provided a port for consul-agent this health check
+      // won't result in it registering with service discovery, but we still
+      // get its health events inside this container
+      name: "consul-agent",
+      exec:  "consul info | grep peers",
+      poll: 5,
+      timeout: "5s"
+    }
+  ]
+  watches: [
+    {
+      name: "database",
+      exec: "reconfigure-db-connection.sh",
```

Comment:
Also, watches will produce events in the form `watch.SomeWatchName`, and will have an explicit field for their service (or, in the future, key) they want to watch.